{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment 4: Using and Sharing Code\n",
    "\n",
    "As usual, be sure to look through the rubric so you know how your work is being evaluated.\n",
    "\n",
    "Also, whenever you start or restart your notebook, run the following code cell to make sure that this notebook pulls in the latest version of your code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Convoluted Plots\n",
    "\n",
    "Imagine that you are collaborating with others to write a computational essay. You have your data and the text of the essay is completely written. However, someone has mixed up the data files, and you have not yet plotted your data. You have been tasked with solving the mystery and plotting the data for your collaborators.\n",
    "\n",
    "In the file `vroom.ipynb`, you will find the current draft of the computational essay.\n",
    "\n",
    "### 1.1. Write a Plotting Function\n",
    "\n",
    "Near the top of `vroom.ipynb`, you will see a function called `make_results_plot` that has been partially filled in. Read the comments in the body of the function, and modify it to appropriately plot a data set from a CSV.\n",
    "\n",
    "In this function, you don't have to modify the docstring, and you should not have to add any new lines of code beyond what is already in the function of the body. Essentially, you only need to use the `plt.plot` function to properly plot the data, and then label the axes and title the plot appropriately.\n",
    "\n",
    "As you complete this function, make sure that you follow the [rules of a good plot](https://softdes.olin.edu/docs/readings/4-using-sharing-code/#data-visualization): your axes should be labeled with a description and units, and the plot should be appropriately titled. You should also make sure that your plot does not use lines to join adjacent points (i.e., each point should be on its own, unconnected to other points).\n",
    "\n",
    "### 1.2. Plot the Correct Data\n",
    "\n",
    "In the `data` folder, you will find files with the names `test_design_1.csv` through `test_design_4.csv`. Four code cells in the notebook have comments for you to fill in with code to plot each of these data sets once.\n",
    "\n",
    "Read the text of the computational essay to determine the characteristics of the data you want to plot. Then, plot the appropriate data set in each cell using the `make_results_plot` function that you wrote in the previous section. You may try plotting all of the data sets and then matching them to the appropriate cells if you prefer."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Gene Finder\n",
    "\n",
    "### 2.0. Introduction\n",
    "\n",
    "Computational approaches for analyzing biological data have revolutionized many subfields of biology. This impact has been so large that a new term has been coined to describe this type of interdisciplinary fusion of biology, computation, math, and statistics: bioinformatics.\n",
    "\n",
    "In this problem, you will be applying your Python and computing skills to a highly relevant present-day problem in bioinformatics: the analysis of the SARS-CoV-2 virus. You are no doubt familiar with this virus as the cause of the COVID-19 pandemic, which has been responsible for likely the most dramatic change to everyday life that we will see in our lifetimes.\n",
    "\n",
    "Specifically, your task is to write code to identify potential protein-coding genes in the SARS-CoV-2 virus, and use the [protein-BLAST](http://www.ncbi.nlm.nih.gov/blast/Blast.cgi?PROGRAM=blastp&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome) search engine to determine which of these genes may encode key proteins of the virus. By doing this, you will get an insight into the process that biologists use when trying to understand something like COVID-19. Specifically, the work you will do in this project roughly corresponds to the \"Gene Prediction\" step in the graphic below.\n",
    "\n",
    "![Genetic Analysis Workflow](gene_finder_workflow.png)\n",
    "\n",
    "Gene prediction is so fundamental for the field in part because once we know where the protein-coding genes are, we can begin to decode the form and function of these proteins, and then the mechanisms that regulate the synthesis of these proteins. With a firm understanding of each of these components of the system, we gain an unprecedented level of understanding and insight into all kinds of biological processes, from infection to symptom development, and hopefully to recovery or treatment.\n",
    "\n",
    "This problem is an opportunity to create a focused, polished piece of software with a real-world application, and thus will require you to implement functions, as well as documentation and tests for each of these functions. It is also an opportunity to create a computational essay of your own, partly in practice for the midterm project in this course. As you work through this problem, we also encourage you to think about the possible uses of this technology. What can this approach to solving problems in biology achieve? What limitations does it have? How can this technology be used to benefit others, or misused to harm others?\n",
    "\n",
    "The problem involves three main components: function implementation, unit tests, and an analytical writeup. We present them in this order, but feel free to use this as an opportunity to try out test-driven development and write the unit tests before your implementation if you wish.\n",
    "\n",
    "### 2.1. Function Implementation\n",
    "\n",
    "The file `gene_finder.py` contains ten functions for you to implement, all of which are described below. You can implement them in any order you choose, but we recommend writing them in the order in which they are described below. If you do so, you will likely find that you can write some functions much more simply by using functions that you have implemented earlier. Additionally, the file `helpers.py` provides some functions and variables that may be helpful, and we expect that you will use each function or variable in this file at least once as you write your functions.\n",
    "\n",
    "Each function description also states the function's assumptions, so be sure to read each description carefully - it will help as you write documentation! If it helps, you can also sneak ahead to `test_gene_finder.py`, which contains a few sample test cases for most of these functions. There, you can see what the function is expected to return for various inputs.\n",
    "\n",
    "#### 2.1.1. `get_complement`\n",
    "\n",
    "The `get_complement` function takes a string `nucleotide` consisting of a single character `A`, `T`, `C`, or `G`, each representing the DNA nucleotides adenine, thymine, cytosine, and guanine, respectively. The function should return a string (also consisting of a single character) representing the complementary nucleotide. In DNA, `A` and `T` are complementary to each other, as are `C` and `G`.\n",
    "\n",
    "#### 2.1.2. `get_reverse_complement`\n",
    "\n",
    "The `get_reverse_complement` function takes a string `strand` representing a single strand of DNA (consisting entirely of the characters `A`, `T`, `C`, and `G`) and returns a string representing a complementary strand of DNA. Complementary strands of DNA run in opposite directions to one another, so this string should be the reverse of simply taking the complement:\n",
    "\n",
    "```\n",
    "------------> Initial strand goes this way\n",
    "ATGCCCGCTTT\n",
    "|||||||||||\n",
    "TACGGGCGAAA\n",
    "<------------ Complementary strand goes this way\n",
    "```\n",
    "\n",
    "Thus in the example above, the reverse complement of `ATGCCCGCTTT` would be `AAAGCGGGCAT` and **not** `TACGGGCGAAA`.\n",
    "\n",
    "#### 2.1.3. `rest_of_orf`\n",
    "\n",
    "An *open reading frame (ORF)* is a sequence of nucleotides in a strand of DNA that can be translated into a protein. An ORF begins with a start codon (the sequence `ATG`) and ends with a stop codon (any of the sequences `TAA`, `TAG`, or `TGA`). Finding these ORFs is a key component of identifying genes of interest in DNA.\n",
    "\n",
    "The `rest_of_orf` function takes a string `strand` representing a strand of DNA that begins with a start codon and returns the sequence of nucleotides representing the rest of the ORF, up to but not including the next stop codon. If there is no stop codon, the function returns the entire strand. Codons in DNA are translated three nucleotides at a time, so the stop codon must appear a multiple of three nucleotides away from the start codon:\n",
    "\n",
    "```\n",
    "ATG ATA GAA TGA\n",
    "```\n",
    "\n",
    "As you can see, `TAG` appears in the sequence earlier (the 5th through 7th nucleotides), but only one nucleotide away from the start codon. The codon `TGA` appears six nucleotides away from the start codon, and is thus counted as a valid stop codon. Thus `rest_of_orf(\"ATGATAGAATGA\")` should return `\"ATGATAGAA\"`.\n",
    "\n",
    "As a reminder, the `amino_acid` function in `helpers.py` can be used to translate nucleotide triplets into their relevant amino acid. The stop codon's symbol is the asterisk (`*`).\n",
    "\n",
    "#### 2.1.4. `find_all_orfs_one_frame`\n",
    "\n",
    "The `find_all_orfs_one_frame` takes a string `strand` representing a strand of DNA and returns a list of strings representing all in-frame ORFs found in that strand. \"In-frame\" means that each ORF should be a multiple of three nucleotides from the start of `strand`. Additionally, you should not include any ORFs completely nested in another, like this:\n",
    "\n",
    "```\n",
    "ATG AAA ATG GCA TGA\n",
    "--------------- Include this\n",
    "        ------- Not this\n",
    "```\n",
    "\n",
    "Assuming that you are scanning the string from left to right, you can avoid having to check for nested ORFs as follows: once you find an ORF, skip to the end of that ORF and continue scanning. A `while` loop would likely be useful for\n",
    "this.\n",
    "\n",
    "As a reminder, the `amino_acid` function in `helpers.py` can be used to translate nucleotide triplets into their relevant amino acid. The start codon's symbol is `M`.\n",
    "\n",
    "#### 2.1.5. `find_all_orfs`\n",
    "\n",
    "The `find_all_orfs` takes a string `strand` representing a strand of DNA and returns a list of strings representing all ORFs found in that strand. This includes not only in-frame ORFs, but also ORFs found one or two nucleotides from the start of `strand`. This means you are effectively scanning triplets of nucleotides like this:\n",
    "\n",
    "```\n",
    "ATG AAA ATG GCA TGA  <- In-frame\n",
    "A TGA AAA TGG CAT GA  <- Frame-shifted by 1 nucleotide\n",
    "AT GAA AAT GGC ATG A  <- Frame-shifted by 2 nucleotides\n",
    "```\n",
    "\n",
    "As with `find_all_orfs_one_frame`, you should avoid nested ORFs, but *only within a single frame*. This means, for example, that if you find an ORF frame-shifted by 1 nucleotide that is completely nested within an ORF found in-frame, you should include both.\n",
    "\n",
    "#### 2.1.6. `find_all_orfs_both_strands`\n",
    "\n",
    "The `find_all_orfs_both_strands` function takes a string `strand` representing a strand of DNA and returns a list of strings representing all ORFs found in `strand` or its reverse complement.\n",
    "\n",
    "#### 2.1.7. `find_longest_orf`\n",
    "\n",
    "The `find_longest_orf` function takes a string `strand` representing a strand of DNA and returns the longest ORF found in either that strand or its reverse complement.\n",
    "\n",
    "#### 2.1.8. `noncoding_orf_threshold`\n",
    "\n",
    "As you find ORFs in a strand of DNA, you may find that many of them are rather short, some being as short as just 3 or 6 nucleotides. These sequences are too short to produce any useful proteins, and should be filtered out.\n",
    "\n",
    "One way to filter these ORFs out is to consider what ORFs you would find in a random DNA strand of this length with the same ratio of nucleotides. You can then take the longest ORF from this random strand and throw out any shorter ORFs from the original strand, with the assumption that those shorter ORFs are likely just due to chance.\n",
    "\n",
    "However, a single random DNA strand may just happen to contain an extremely long ORF by chance, leading many of your ORFs from the original strand to be thrown out. To prevent this, you should consider many random DNA strands and find the \"shortest longest\" ORF (the shortest ORF out of the longest ORFs from all of the random strands you considered).\n",
    "\n",
    "The `noncoding_orf_threshold` takes a string `strand` representing a strand of DNA and a positive integer `num_trials` representing a number of trials to run. For each of these trials, it randomly shuffles the nucleotides in `strand` and\n",
    "finds the longest ORF in either the shuffled strand or its reverse complement. It then keeps track of the *minimum* length of this value over all of the trials and returns an integer representing this minimum length. As an example, if the function does 5 trials and finds longest ORFs of lengths 444, 294, 429, 243, and 264 in the trials, then the function should return 243.\n",
    "\n",
    "#### 2.1.9. `encode_amino_acids`\n",
    "\n",
    "Once you have identified a set of candidate ORFs of suitable length, you should determine the sequence of amino acids these ORFs encode. Each of these sequences is potentially a protein of interest in the genome, and will be a candidate for further analysis. You will start by encoding a single ORF into a sequence of amino acids.\n",
    "\n",
    "The `encode_amino_acids` function takes a string `orf` representing a strand of DNA that is an ORF and returns a string representing the sequence of amino acids, with each amino acid written as its one-letter symbol. The symbols for each amino acid are defined by the [International Union of Pure and Applied Chemistry (IUPAC)](http://iupac.org/), and the mapping between codons and the amino acid symbol can be found [on Wikipedia](https://en.wikipedia.org/wiki/DNA_codon_table). There is also a convenience function in `helpers.py` that you can use.\n",
    "\n",
    "It is possible that an ORF's length will not be an exact multiple of 3, due to reading an ORF from an offset. In this case, there will be 1 or 2 nucleotides left at the end - your implementation should simply ignore these.\n",
    "\n",
    "#### 2.1.10. `find_genes`\n",
    "\n",
    "Finally, it's time to put everything together and find potential protein-coding genes. The `find_genes` function takes a string `path` representing the location of a file in [FASTA\n",
    "format](https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=BlastHelp). It should do the following:\n",
    "\n",
    "1. Load the string of nucleotides from the file.\n",
    "2. Determine the threshold length to use as a cutoff for coding ORFs, using 1,500 trials of shuffling the nucleotide sequence.\n",
    "3. Find all ORFs in both strands of the sequence.\n",
    "4. For any ORF longer than the cutoff length, translate the ORF to a sequence of amino acids.\n",
    "5. Return the list of all such amino acid sequences.\n",
    "\n",
    "As a reminder, this function should heavily utilize other functions from `helpers.py` and `gene_finder.py`.\n",
    "\n",
    "Once you have completed all of the above steps, your implementation is complete!\n",
    "\n",
    "### 2.2. Unit Tests\n",
    "\n",
    "To make sure that your submitted implementation works properly, you should write a thorough set of unit tests. In the `test_gene_finder.py` file, you will find some testing code to start from. The starter code provided in this file contains all of the code needed to perform some initial tests on your code, but you will need to write additional test cases.\n",
    "\n",
    "The testing file contains a number of lists with names like `get_complement_cases`. Each list contains a series of tuples, each with two items. The name of the list up to `_cases` indicates which function these test cases are for, and the two items in each tuple are a test input and the expected output, respectively. The existing cases have a comment above them indicating what the unit test is specifically testing for.\n",
    "\n",
    "For each of the test case lists, add test cases to thoroughly test the function it corresponds to. Each test case **must** have a comment to explain what the test case is checking or why you added that test. While for some functions\n",
    "adding one or two test cases will suffice, for others more may be necessary. **Your tests should show that you have carefully thought about the possible behavior of the function and made sure that it returns the correct value in a variety of scenarios.**\n",
    "\n",
    "You may notice that in the lists, even the last test case's tuple has a comma (`,`) after it - this is fine, and makes it easy to reorganize test cases without having to add/delete commas in the list.\n",
    "\n",
    "**You should consider writing additional test functions where helpful.** An example of this is the `test_double_complement` function that we have provided for you, which checks that taking the complement of a complement of a nucleotide returns the original nucleotide. While you do not have to write additional test functions for every function in `gene_finder.py`, you may find it helpful to do so to thoroughly test your functions.\n",
    "\n",
    "Note that two functions do not have unit tests for you to write: `noncoding_orf_length` and `find_genes`. This is because both of these functions rely on randomization and thus it is difficult to write unit tests for them that will produce a consistent result.\n",
    "\n",
    "**Extra credit**: when you have submitted your assignment, we will run a collection of all submitted tests in addition to a few we have written. For each test that you write that correctly catches an incorrect implementation where our tests do not, you will receive 1 extra credit point, up to a maximum of 10 points.\n",
    "\n",
    "### 2.3. Analysis and Writeup\n",
    "\n",
    "Once you have finished your implementation and tests, you should use your code to help you analyze the virus's genetic material and write up your findings as a [computational essay](https://softdes.olin.edu/docs/readings/4-using-sharing-code/#computational-essays).\n",
    "\n",
    "In the file `sars-cov-2.ipynb`, you will find the beginnings of a computational essay for this portion of the problem. In this essay, write up a description of how you used your code to identify possible protein-coding genes and how you matched these genes to known viral proteins (see below for a more detailed description on this).\n",
    "\n",
    "Your computational essay should clearly show the computations you ran to get your results (i.e., it needs to include code cells). It should also tell a compelling story - in other words, simply providing the information asked for without connecting them to the overall context is not sufficient. The ideal essay will describe what code you ran, how you then analyzed those results, and an interpretation of those results, at a minimum. (Remember that the notebook in `vroom.ipynb` and Stephen Wolfram's rather meta [computational essay on what a computational essay is](https://writings.stephenwolfram.com/2017/11/what-is-a-computational-essay/) are good examples of how a computational essay should flow.) It may help for you to create your own sections as appropriate.\n",
    "\n",
    "Your essay should describe the use of your code to perform a starting analysis of the SARS-CoV-2 genome. In the code portion of this analysis, you should import the necessary code from `gene_finder.py` and use it to read and analyze a file containing the genetic data of SARS-CoV-2. For the purposes of this project, the file is in `data/NC_045512.2.fa`, which contains the nucleotide sequence of the SARS-CoV-2 virus (specifically, the isolate Wuhan-Hu-1 strain of the virus, which was the first to be sequenced).\n",
    "\n",
    "Note that running the main `find_genes` function may take some time, as performing the necessary shuffles to determine the noncoding ORF threshold length can be a bit slower than the usual functions we have been writing in this course.\n",
    "\n",
    "Once you have identified the candidate genes (which should be shown in the output of your code cell), you should look up these genes in [protein-BLAST](http://www.ncbi.nlm.nih.gov/blast/Blast.cgi?PROGRAM=blastp&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome) to find information about five major proteins in the DNA sequence. These proteins are all significant because having a better understanding of these proteins can give us insights into how the virus works, as well as possible routes to treatment or a vaccine. Here is a brief description of each of the proteins for you to find, along with a brief summary of the significance of each protein:\n",
    "\n",
    "* Polyprotein ORF1a (or ORF1ab in some cases) is a large \"precursor protein\" from which many of the virus's proteins are formed. This is done by cutting or cleaving the polyprotein at certain locations.\n",
    "* The nucleocapsid protein carries the virus's genetic material.\n",
    "* The envelope protein forms part of the outer layer of the virus, and protects it from the host's immune system as the virus travels between host cells.\n",
    "* The membrane protein is also part of the outer layer of the virus, and fuses with the host cell's membrane (the cell's outer layer) when the virus enters the host cell.\n",
    "* The spike protein binds with specific receptors on the host cell, starting the infection of the host cell. The spike protein ensures that the virus only infects the type(s) of host cells that it is suited for.\n",
    "\n",
    "As you might imagine, these proteins are interesting because treatment or vaccination could potentially target them - for example, treatment might hinder the spike protein's ability to bind with and infect the proper host cell, reducing the virus's chances of infecting a person.\n",
    "\n",
    "To find each of the proteins above, do the following:\n",
    "\n",
    "* Look up some of the genes you found in protein-BLAST. To do this, you can copy and paste the amino acid sequence (beginning with `M`) into the Query Sequence box. Note that for some sequences, the search can take a while (but usually no more than a couple minutes).\n",
    "* The search results will indicate a list of potential matches. By reading the descriptions, you should be able to determine whether or not the gene matches one of the five proteins described above. For each gene matching one of the five proteins, note the accession number (the alphanumeric string in the rightmost column of the search results) and the number of amino acids in the protein (the number following \"Query Length\" above the search results).\n",
    "* Include these results in your written analysis.\n",
    "\n",
    "As you look through the genes you identified with your code, you may find that some of these genes do not match any of the proteins. Feel free to look these up and write about them if you are so inclined, though this is completely optional.\n",
    "\n",
    "#### 2.4. Reflection\n",
    "\n",
    "If you have completed the above three items, you have gone through nearly all of the steps that we expect you to complete in the midterm project. However, doing computing is not just about completing a technical task; as a responsible engineer, you should also be reflecting on the work that you do and the contextual and ethical implications of your work.\n",
    "\n",
    "The file `reflection.md` contains two labeled sections in which you should write a paragraph to answer each of the following:\n",
    "\n",
    "* What are some limitations of the code you wrote? Are there genes or proteins that you would not be able to identify using your code? (See [this visual representation of the SARS-CoV-2 genome](https://www.ncbi.nlm.nih.gov/projects/sviewer/?id=NC_045512.2) to get a better sense of the genes in the virus.) In what ways could you adapt or extend your code to address these limitations?\n",
    "* Hopefully, this project has allowed you to see how gene finding has the potential to benefit others. What are some other ways besides analyzing viral or bacterial DNA in which identifying genes and proteins can benefit others? How might these techniques be used to *negatively* affect others, and what are some possible ways to address these risks?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Acknowledgments\n",
    "\n",
    "Thie gene finder problem was originally created by Professors Ran Libeskind-Hadas, Eliot C. Bush, and their collaborators at [Harvey Mudd College](https://www.cs.hmc.edu/twiki/bin/view/CS6/GreenAcknowledgements). Special thanks to Ran for allowing us to use this assignment and adapt it for this course. Thanks also to the Summer 2019 Context and Ethics team for updates to the problem description."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}